compiler/test/src/test/java/org/apache/nemo/compiler/frontend/beam/transform/GroupByKeyAndWindowDoFnTransformTest.java

```diff
@@ -99,76 +106,148 @@ public void test() {
 
     doFnTransform.onData(WindowedValue.of(
-      KV.of("2", "hello"), ts3, fixedwindows.assignWindow(ts3), PaneInfo.NO_FIRING));
+      KV.of("2", "hello"), ts3, slidingWindows.assignWindows(ts3), PaneInfo.NO_FIRING));
 
     doFnTransform.onWatermark(watermark);
 
+    // output
+    // 1: ["hello", "world"]
+    // 2: ["hello"]
     Collections.sort(oc.outputs, (o1, o2) -> o1.getValue().getKey().compareTo(o2.getValue().getKey()));
 
     // windowed result for key 1
-    assertEquals(Arrays.asList(fixedwindows.assignWindow(ts1)), oc.outputs.get(0).getWindows());
+    assertEquals(Arrays.asList(window0), oc.outputs.get(0).getWindows());
     checkOutput(KV.of("1", Arrays.asList("hello", "world")), oc.outputs.get(0).getValue());
 
     // windowed result for key 2
-    assertEquals(Arrays.asList(fixedwindows.assignWindow(ts3)), oc.outputs.get(1).getWindows());
+    assertEquals(Arrays.asList(window0), oc.outputs.get(1).getWindows());
     checkOutput(KV.of("2", Arrays.asList("hello")), oc.outputs.get(1).getValue());
 
+    assertEquals(2, oc.outputs.size());
+    assertEquals(1, oc.watermarks.size());
+
     // check output watermark
-    assertEquals(fixedwindows.assignWindow(ts1).maxTimestamp().getMillis(),
+    assertEquals(1000,
       oc.watermarks.get(0).getTimestamp());
 
     oc.outputs.clear();
     oc.watermarks.clear();
 
     doFnTransform.onData(WindowedValue.of(
-      KV.of("1", "a"), ts4, fixedwindows.assignWindow(ts4), PaneInfo.NO_FIRING));
+      KV.of("1", "a"), ts4, slidingWindows.assignWindows(ts4), PaneInfo.NO_FIRING));
 
     // do not emit anything
     doFnTransform.onWatermark(watermark2);
     assertEquals(0, oc.outputs.size());
     assertEquals(0, oc.watermarks.size());
 
     doFnTransform.onData(WindowedValue.of(
-      KV.of("2", "a"), ts5, fixedwindows.assignWindow(ts5), PaneInfo.NO_FIRING));
+      KV.of("3", "a"), ts5, slidingWindows.assignWindows(ts5), PaneInfo.NO_FIRING));
 
     doFnTransform.onData(WindowedValue.of(
-      KV.of("3", "a"), ts6, fixedwindows.assignWindow(ts6), PaneInfo.NO_FIRING));
+      KV.of("3", "a"), ts6, slidingWindows.assignWindows(ts6), PaneInfo.NO_FIRING));
 
     doFnTransform.onData(WindowedValue.of(
-      KV.of("2", "b"), ts7, fixedwindows.assignWindow(ts7), PaneInfo.NO_FIRING));
+      KV.of("3", "b"), ts7, slidingWindows.assignWindows(ts7), PaneInfo.NO_FIRING));
 
-    // emit windowed value
+    // emit window1
     doFnTransform.onWatermark(watermark3);
 
+    // output
+    // 1: ["hello", "world", "a"]
+    // 2: ["hello"]
+    // 3: ["a", "a", "b"]
     Collections.sort(oc.outputs, (o1, o2) -> o1.getValue().getKey().compareTo(o2.getValue().getKey()));
 
+
     // windowed result for key 1
-    assertEquals(Arrays.asList(fixedwindows.assignWindow(ts4)), oc.outputs.get(0).getWindows());
-    checkOutput(KV.of("1", Arrays.asList("a")), oc.outputs.get(0).getValue());
+    assertEquals(Arrays.asList(window1), oc.outputs.get(0).getWindows());
```

Comment:
Can you also check for the `PaneInfo` of the output elements?
(`EARLY`, `ON_TIME`, and `LATE`).

It'd be definitely nice to experiment with different triggers in the `WindowingStrategy` to ensure correct behaviors of the watermark holds, which I suppose allows the handling of `EARLY` and `LATE` elements, in addition to the timers mostly handling `ON_TIME`.